compiler/qsc_eval/src/intrinsic/utils.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use std::collections::hash_map::Entry;
+
+use num_bigint::BigUint;
+use num_complex::{Complex, Complex64};
+use num_traits::{One, Zero};
+use rustc_hash::FxHashMap;
+
+pub fn split_state(
+    qubits: &[usize],
+    state: Vec<(BigUint, Complex64)>,
+    qubit_count: usize,
+) -> Option<Vec<(BigUint, Complex64)>> {
```

Comment:
I think returning `Result<Vec<(BigUint, Complex64)>, ()>` would be more appropriate since it conveys the fact that an error happens when trying to split a non-separable state.

Reply:
And it would also make it easier to quickly see the cases where this happens.